### DISCUSSION TITLE
Help deploying conductor to heroku

### DISCUSSION THREAD

**Discussion by roopeshshettyb** [Netflix/conductor#3013] [Q&A]
Hi,

I have been using Conductor in my project locally, I wanted to deploy it to heroku so I could have a production version.

I cloned the repository and tried deploying with heroku but I get the following error ->
```
FAILURE: Build failed with an exception.
remote:
remote:        * Where:
remote:        Build file '/tmp/build_3f53ba5a/build.gradle' line: 22
remote:
remote:        * What went wrong:
remote:        An exception occurred applying plugin request [id: 'nebula.netflixoss', version: '10.6.0']
remote:        > Failed to apply plugin class 'nebula.plugin.netflixossproject.publishing.PublishingPlugin'.
remote:           > Task with name 'postRelease' not found in root project 'conductor'.
```

Could someone please help me with this? 
I just cloned the repo and tried to deploy, I didn't make any changes to the code and I am using the latest version.

Full log ->

> remote: Building source:
> remote: 
> remote: -----> Building on the Heroku-20 stack
> remote: -----> Determining which buildpack to use for this app
> remote: -----> Gradle app detected
> remote: -----> Spring Boot detected
> remote: -----> Installing JDK 1.8... done
> remote: -----> Building Gradle app...
> remote: -----> executing ./gradlew build -x check
> remote:        Downloading https://services.gradle.org/distributions/gradle-7.4.2-bin.zip
> remote:        ...........10%...........20%...........30%...........40%...........50%...........60%...........70%...........80%...........90%...........100%
> remote:        To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.4.2/userguide/gradle_daemon.html#sec:disabling_the_daemon.
> remote:        Daemon will be stopped at the end of the build 
> remote:        
> remote:        > Configure project :
> remote:        Git repository not found at /tmp/build_3f53ba5a -- nebula-release tasks will not be available. Use the git.root Gradle property to specify a different directory.
> remote:        
> remote:        FAILURE: Build failed with an exception.
> remote:
> remote:        * Where:
> remote:        Build file '/tmp/build_3f53ba5a/build.gradle' line: 22
> remote:
> remote:        * What went wrong:
> remote:        An exception occurred applying plugin request [id: 'nebula.netflixoss', version: '10.6.0']
> remote:        > Failed to apply plugin class 'nebula.plugin.netflixossproject.publishing.PublishingPlugin'.
> remote:           > Task with name 'postRelease' not found in root project 'conductor'.
> remote:
> remote:        * Try:
> remote:        > Run with --stacktrace option to get the stack trace.
> remote:        > Run with --info or --debug option to get more log output.
> remote:        > Run with --scan to get full insights.
> remote:
> remote:        * Get more help at https://help.gradle.org
> remote:  !     ERROR: Failed to run Gradle!

**Comment by v1r3n** [comment]
I see its using JDK1.8  `remote: -----> Installing JDK 1.8... done` 
Conductor needs JDK 11, can you try with JDK11?

**Reply by roopeshshettyb** [reply]
Thank you for your reply, heroku does 1.8 by default so I made a config change so it uses jdk11. 
Unfortunately that didnt fix the issue. Still the same error, but this time it did say it installed jdk 11

**Reply by aravindanr** [reply]
Can you run the build with [debugging enabled](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:command_line_debugging) and paste the stack trace?

`./gradlew build -s -x check`

**Reply by roopeshshettyb** [reply]
There's no way I could find to do that with heroku. I tried the heroku bash CLI, but could not trigger the build from there. The build can only be triggered by pushing to the heroku git afaik. After the push is successful, it auto deploys.

I ran that command locally on my PS and got a build successful msg

```
PS R:\conductor> ./gradlew build -s -x check

> Configure project :
Inferred project: conductor, version: 3.9.0-SNAPSHOT

> Task :conductor-common:protogen
Running protogen with arguments: ProtoGenTask{protoPackage='conductor.proto', javaPackage='com.netflix.conductor.proto', goPackage='github.com/netflix/conductor/client/gogrpc/conductor/model', protosDir=R:\conductor\grpc\src\main\proto, mapperDir=R:\conductor\grpc\src\main\java\com\netflix\conductor\grpc, mapperPackage='com.netflix.conductor.grpc', sourceJar=R:\conductor\common\build\libs\conductor-common-3.9.0-SNAPSHOT.jar, sourcePackage='com.netflix.conductor.common'}
protogen: processing Jar 'R:\conductor\common\build\libs\conductor-common-3.9.0-SNAPSHOT.jar'
protogen: found com.netflix.conductor.common.metadata.events.EventExecution
protogen: found com.netflix.conductor.common.metadata.events.EventHandler
protogen: found com.netflix.conductor.common.metadata.tasks.PollData
protogen: found com.netflix.conductor.common.metadata.tasks.Task
protogen: found com.netflix.conductor.common.metadata.tasks.TaskDef
protogen: found com.netflix.conductor.common.metadata.tasks.TaskExecLog
protogen: found com.netflix.conductor.common.metadata.tasks.TaskResult
protogen: found com.netflix.conductor.common.metadata.workflow.DynamicForkJoinTask
protogen: found com.netflix.conductor.common.metadata.workflow.DynamicForkJoinTaskList
protogen: found com.netflix.conductor.common.metadata.workflow.RerunWorkflowRequest
protogen: found com.netflix.conductor.common.metadata.workflow.SkipTaskRequest
protogen: found com.netflix.conductor.common.metadata.workflow.StartWorkflowRequest
protogen: found com.netflix.conductor.common.metadata.workflow.SubWorkflowParams
protogen: found com.netflix.conductor.common.metadata.workflow.WorkflowDef
protogen: found com.netflix.conductor.common.metadata.workflow.WorkflowTask
protogen: found com.netflix.conductor.common.run.TaskSummary
protogen: found com.netflix.conductor.common.run.Workflow
protogen: found com.netflix.conductor.common.run.WorkflowSummary
protogen: writing 'R:\conductor\grpc\src\main\java\com\netflix\conductor\grpc\AbstractProtoMapper.java'...
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/C:/Users/Roopesh/.gradle/caches/7.4.2/generated-gradle-jars/gradle-api-7.4.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/C:/Users/Roopesh/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-slf4j-impl/2.17.2/183f7c95fc981f3e97d008b363341343508848e/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.gradle.internal.logging.slf4j.OutputEventListenerBackedLoggerContext]
protogen: writing 'R:\conductor\grpc\src\main\proto\model\dynamicforkjointask.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\dynamicforkjointasklist.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\eventexecution.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\eventhandler.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\polldata.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\rerunworkflowrequest.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\skiptaskrequest.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\startworkflowrequest.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\subworkflowparams.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\task.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\taskdef.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\taskexeclog.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\taskresult.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\tasksummary.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\workflow.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\workflowdef.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\workflowsummary.proto'...
protogen: writing 'R:\conductor\grpc\src\main\proto\model\workflowtask.proto'...
protogen completed.

> Task :conductor-server:compileJava
Execution optimizations have been disabled for task ':conductor-server:compileJava' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'R:\conductor\server\build\resources\main'. Reason: Task ':conductor-server:compileJava' uses this output of task ':conductor-server:bootBuildInfo' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
Note: R:\conductor\server\src\main\java\com\netflix\conductor\Conductor.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4.2/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD SUCCESSFUL in 22s
157 actionable tasks: 6 executed, 151 up-to-date

Publishing build scan...
https://gradle.com/s/mdknbwdo27wug
```

**Reply by apanicker-nflx** [reply]
Please ensure that you are using the gradle version alongwith the server when you build. Older gradle versions may not be compatible with the new plugins.

**Comment by matteius** [comment]
I am getting the same issue trying to deploy to Heroku and I have updated the config to use OpenJDK 11 -- it works locally.

**Reply by apanicker-nflx** [reply]
Please ensure that you are using the gradle version alongwith the server when you build. Older gradle versions may not be compatible with the new plugins.
